linkerd/proxy/pool/src/worker.rs

```diff
@@ -0,0 +1,349 @@
+                        terminate
+                            .take()
+                            .expect("must not fail twice")
+                            .send(err.clone());
+                        reqs_rx.close();
+                        terminal_failure = Some(err);
+                        tracing::trace!("Closed");
+                    } else {
+                        tracing::trace!("Pool ready");
+                    }
+                }
+
+                // Process requests, either by dispatching them to the pool or
+                // by serving errors directly.
+                let _ = if let Some(e) = terminal_failure.clone() {
+                    tx.send(Err(e.into()))
+                } else {
+                    tx.send(worker.pool.call(req))
+                };
+
+                // TODO(ver) track histogram from t0 until the request is dispatched.
+                tracing::debug!(
+                    latency = (time::Instant::now() - t0).as_secs_f64(),
+                    "Dispatched"
+                );
+            }
+        }
+        .instrument(debug_span!("pool")),
+    )
+}
+
+// === impl Worker ===
+
+impl<T, R, P> Worker<R, P>
+where
+    T: Clone + Eq + std::fmt::Debug,
+    R: TryStream<Ok = Update<T>> + Unpin,
+    R::Error: Into<Error>,
+{
+    /// Attempts to update the pool with discovery updates.
+    ///
+    /// Additionally, this attempts to drive the pool to ready if it is
+    /// currently in failfast.
+    ///
+    /// If the discovery stream is closed, this never returns.
+    async fn discover_while_awaiting_requests<Req>(&mut self) -> Error
+    where
+        P: Pool<T, Req>,
+        P::Error: Into<Error>,
+    {
+        tracing::trace!("Discovering while awaiting requests");
+
+        loop {
+            let update = tokio::select! {
+                e = self.pool.drive() => return e,
+                res = self.discovery.discover() => match res {
+                    Err(e) => return e,
+                    Ok(up) => up,
+                },
+            };
+
+            tracing::debug!(?update, "Discovered");
+            self.pool.pool.update_pool(update);
+        }
+    }
+
+    /// Wait for the pool to have at least one ready endpoint, while also
+    /// processing service discovery updates (e.g. to provide new available
+    /// endpoints).
+    async fn ready_pool<Req>(&mut self) -> Result<(), Error>
+    where
+        P: Pool<T, Req>,
+        P::Error: Into<Error>,
+    {
+        loop {
+            tokio::select! {
+                // Tests, especially, depend on discovery updates being
+                // processed before ready returning.
+                biased;
+
+                // If the pool updated, continue waiting for the pool to be
+                // ready.
+                res = self.discovery.discover() => {
+                    let update = res?;
+                    tracing::debug!(?update, "Discovered");
```

Comment:
take it or leave it: maybe worth swapping the `let` and the trace so that we also log on errors?
```suggestion
                update = self.discovery.discover() => {
                    tracing::debug!(?update, "Discovered");
                    let update = update?;
```